File: java/Code23.java
import java.io.*;
import java.util.*;
public class Code23 {
  static ArrayList<Integer>[] A;
  static boolean visited[];
  @SuppressWarnings("unchecked")
public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    int n = Integer.parseInt(st.nextToken());
    int m = Integer.parseInt(st.nextToken());
    A = new ArrayList[n + 1];
    visited = new boolean[n + 1];
    for (int i = 1; i < n + 1; i++) { // 인접 리스트 초기화
      A[i] = new ArrayList<Integer>();
    }
    for (int i = 0; i < m; i++) {
      st = new StringTokenizer(br.readLine());
      int s = Integer.parseInt(st.nextToken());
      int e = Integer.parseInt(st.nextToken());
      A[s].add(e); // 양 방향 간선이므로 양쪽으로 간선을 더 해준다
      A[e].add(s);
    }
    int count = 0;
    for (int i = 1; i < n + 1; i++) {
      if (!visited[i]) { // 미 방문한 정점이 없을 때까지 반복
        count++;
        DFS(i);
      }
    }
    System.out.println(count);
  }
  static void DFS(int v) {
    if (visited[v]) {
      return;
    }
    visited[v] = true;
    for (int i : A[v]) {
      if (visited[i] == false) { // 연결 정점 중 방문하지 않았던 정점만 탐색함
        DFS(i);
      }
    }
  }
}

/*  @SuppressWarnings("unchecked")
  자바에서는 [] 배열에서 업케스팅 다운케스팅이 빈번하게 발생하게 되는데
 
제네릭은 캐스팅이 불가능합니다. 따라서 거기에 대해서 100% 안전하게 타입을 거를 수 없으니까,

프로그래머가 확실히 맞는타입만 올 것이라고 저렇게 선언해주는 수 밖에 없습니다.*/

File: python/Code23.py
import sys
sys.setrecursionlimit(10000)
input = sys.stdin.readline
n, m = map(int, input().split())
A = [[] for _ in range(n+1)]
visited = [False] * (n+1)

def DFS(v):
    visited[v] = True
    for i in A[v]:
        if not visited[i]:
            DFS(i)

for _ in range(m):
    s, e = map(int, input().split())
    A[s].append(e)  # 양방향 에지이므로 양쪽에 에지를 더하기
    A[e].append(s)
count = 0
for i in range(1, n+1):
    if not visited[i]:  # 연결 노드 중 방문하지 않았던 노드만 탐색
        count += 1
        DFS(i)
print(count)

#파이썬으로 재귀를 사용해 문제를 풀 때 특히, DFS, BFS 문제를 풀 때 예시에서 답은 잘 나오는데,
#  정답 제출을 하면 런타임 에러를 접하게 되는 경우가 있습니다.
#대부분이 파이썬의 재귀 최대 깊이의 기본 설정이 1,000회 이기 때문에 런타임 에러가 발생하는 경우입니다.
#이런 문제를 해결하기 위해서는 아래와 같이 코드를 작성해주면 됩니다.
#sys.setrecursionlimit(10 ** 6)

#A =[[],[2,5],[1,5],[4],[3,6],[1,2],[4]]
# 첫번째 재귀 visited :[False, True , True, False, False, True, False]
# 두번째 재귀 visited :[False, True , True, True, True, True, True]